### PR TITLE
Convert `user-features` to TypeScript

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Convert user-features to TypeScript. This process was done looking at the existing flow types, but the following changes have been introduced:
- cookie interaction using `@guardian/libs`
- a validation step for the JSON response
- merging the missing cookie check with an old cookie check
- in the test, we mock a valid response from the API

The question remains as to whether we can memoize some of the cookie calls, but that might be a premature optimisation.

cc @guardian/reader-revenue for visibility.

## Does this change need to be reproduced in dotcom-rendering ?

- [X] No
- [ ] Yes (please indicate your plans for DCR Implementation)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [X] No
- [ ] Yes

### Does this change break ad-free?

- [X] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/commercial/user-features.ts [hunk 1]
```diff
@@ -35,124 +39,167 @@ const CONTRIBUTIONS_REMINDER_SIGNED_UP = {
...
+		isObject(response.contentAccess) &&
+		typeof response.contentAccess.paidMember === 'boolean' &&
+		typeof response.contentAccess.recurringContributor === 'boolean' &&
+		typeof response.contentAccess.digitalPack === 'boolean'
+	) {
+		return true;
+	}
+
+	return true;
+};
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/commercial/user-features.ts:92]: This is a new validation step for the incoming object.

File: static/src/javascripts/projects/common/modules/commercial/user-features.ts [hunk 2]
```diff
@@ -35,124 +39,167 @@ const CONTRIBUTIONS_REMINDER_SIGNED_UP = {
...
+		{
+			mode: 'cors',
+			credentials: 'include',
+		},
+	)
+		.then((response) => {
+			if (!validateResponse(response))
+				throw new Error('invalid response');
+			return response;
+		})
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/commercial/user-features.ts:170]: We now have a validation step, which ensures our response is valid.

File: static/src/javascripts/projects/common/modules/commercial/user-features.ts [hunk 3]
```diff
@@ -35,124 +39,167 @@ const CONTRIBUTIONS_REMINDER_SIGNED_UP = {
...
-	const expiryDateFromCookie = getCookie(datedCookieName);
+const datedCookieIsOldOrMissing = (datedCookieName: string): boolean => {
+	const expiryDateFromCookie = getCookie({ name: datedCookieName });
+	if (!expiryDateFromCookie) return true;
 	const expiryTime = parseInt(expiryDateFromCookie, 10);
 	const timeNow = new Date().getTime();
 	return timeNow >= expiryTime;
 };
 
-const featuresDataIsMissing = () => !getCookie(USER_FEATURES_EXPIRY_COOKIE);
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/commercial/user-features.ts:133]: Removed `featuresDataIsMissing` and merged with `featuresDataIsOld`.

**Review by mxdvl — COMMENTED** [review]
Bits of interests highlighted below 👇

File: static/src/javascripts/projects/common/modules/commercial/user-features.ts [hunk 4]
```diff
@@ -35,124 +39,167 @@ const CONTRIBUTIONS_REMINDER_SIGNED_UP = {
...
+		})
 		.then(persistResponse)
-		.catch(() => {});
+		.catch(noop);
 
-const datedCookieIsOld = (datedCookieName) => {
-	const expiryDateFromCookie = getCookie(datedCookieName);
+const datedCookieIsOldOrMissing = (datedCookieName: string): boolean => {
+	const expiryDateFromCookie = getCookie({ name: datedCookieName });
+	if (!expiryDateFromCookie) return true;
```
Comment (arelra) [inline static/src/javascripts/projects/common/modules/commercial/user-features.ts:176]: Minor non-blocking - just some naming improvements if you agree:
```suggestion
const cookieIsExpiredOrMissing = (cookieName: string): boolean => {
	const expiryDateFromCookie = getCookie({ name: cookieName });
	if (!expiryDateFromCookie) return true;
```

File: static/src/javascripts/projects/common/modules/commercial/user-features.ts [hunk 5]
```diff
@@ -35,124 +39,167 @@ const CONTRIBUTIONS_REMINDER_SIGNED_UP = {
...
 	const expiryTime = parseInt(expiryDateFromCookie, 10);
 	const timeNow = new Date().getTime();
 	return timeNow >= expiryTime;
 };
 
-const featuresDataIsMissing = () => !getCookie(USER_FEATURES_EXPIRY_COOKIE);
+const featuresDataIsOld = () =>
+	datedCookieIsOldOrMissing(USER_FEATURES_EXPIRY_COOKIE);
```
Comment (arelra) [inline static/src/javascripts/projects/common/modules/commercial/user-features.ts:183]:
```suggestion
	cookieIsExpiredOrMissing(USER_FEATURES_EXPIRY_COOKIE);
```

File: static/src/javascripts/projects/common/modules/commercial/user-features.ts [hunk 6]
```diff
@@ -35,124 +39,167 @@ const CONTRIBUTIONS_REMINDER_SIGNED_UP = {
...
-
-const adFreeDataIsOld = () => {
-	const switches = config.get('switches');
+const adFreeDataIsOld = (): boolean => {
+	const { switches } = window.guardian.config;
 	return (
-		switches.adFreeStrictExpiryEnforcement &&
-		datedCookieIsOld(AD_FREE_USER_COOKIE)
+		Boolean(switches.adFreeStrictExpiryEnforcement) &&
+		datedCookieIsOldOrMissing(AD_FREE_USER_COOKIE)
```
Comment (arelra) [inline static/src/javascripts/projects/common/modules/commercial/user-features.ts:189]:
```suggestion
		cookieIsExpiredOrMissing(AD_FREE_USER_COOKIE)
```

**Review by arelra — APPROVED** [review]

**Pull request merged**